## 6단계: 수동 역전파

> 이전 단계에서 역전파의 구동 원리를 설명했습니다. \
이번 단계에서는 Variable과 Function 클래스를 확장하여 역전파를 이용한 미분을 구현하겠습니다. \
Variable 클래스부터 살펴보죠.

### 6.1 Variable 클래스 추가 구현

통상값(data)과 더불어 그에 대응하는 미분값(grad)도 저장하도록 확장한다.
- grad: gradient(그래디언트)의 약어

In [3]:
import numpy as np

In [1]:
class Variable:
    def __init__(self, data):
        self.data = data
        self.grad = None  # 추후 역전파 하면 미분값 계산하여 대입

### 6.2 Function 클래스 추가 구현

다음 두 기능 추가
- 미분을 계산하는 역전파(backward 메서드)
- forward 메서드 호출 시 건네받은 Variable 인스턴스 유지

In [2]:
class Function:
    def __call__(self, input):
        x = input.data
        y = self.forward(x)
        output = Variable(y)
        self.input = input  # 입력 변수를 기억(보관)한다.
        return output
    
    def forward(self, x):
        raise NotImplementedError
    
    def backward(self, gy):
        # backward 메서드 구현
        raise NotImplementedError

### 6.3 Square와 Exp 클래스 추가 구현

$y=x^2$의 미분은 $\frac{dy}{dx}=2x$

$y=e^x$의 미분은 $\frac{dy}{dx}=e^x$

In [4]:
class Square(Function):
    def forward(self, x):
        y = x ** 2
        return y
    
    def backward(self, gy):
        x = self.input.data
        gx = 2 * x * gy
        return gx


class Exp(Function):
    def forward(self, x):
        y = np.exp(x)
        return y
    
    def backward(self, gy):
        x = self.input.data
        gx = np.exp(x) * gy
        return gx

### 6.4 역전파 구현

순전파 계산

<img src="images/그림 6-1.png" width=540/>

In [5]:
A = Square()
B = Exp()
C = Square()

x = Variable(np.array(0.5))
a = A(x)
b = B(a)
y = C(b)

역전파 계산

<img src="images/그림 6-2.png" width=640/>

In [6]:
y.grad = np.array(1.0)
b.grad = C.backward(y.grad)
a.grad = B.backward(b.grad)
x.grad = A.backward(a.grad)
print(x.grad)

3.297442541400256


역전파의 계산 결과가 4단계의 해석적 미분 계산결과와 동일함을 알 수 있다.

(해석적 미분 결과: 3.297442541400256)